goodgopher/goodgopher.go

```diff
@@ -0,0 +1,249 @@
+	"github.com/sirupsen/logrus"
+	git "gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+type server struct {
+	appID int
+	key   []byte
+}
+
+// New returns an http.Handler that provides the entrypoints needed for goodgopher.
+func New(appID int, key []byte) (http.Handler, error) {
+	return &server{appID, key}, nil
+}
+
+func (s *server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	kind := r.Header.Get("X-Github-Event")
+	logrus.Debugf("received hook of kind %s", kind)
+	switch kind {
+	case "pull_request":
+		var data github.PullRequestEvent
+		if err := json.NewDecoder(r.Body).Decode(&data); err != nil {
+			logrus.Errorf("could not decode request: %v", err)
+			http.Error(w, "could not decode request", http.StatusInternalServerError)
+			return
+		}
+
+		inst := data.GetInstallation()
+		t, err := ghinstallation.New(http.DefaultTransport, s.appID, int(inst.GetID()), s.key)
+		if err != nil {
+			logrus.Error(err)
+			http.Error(w, "could not authenticate with GitHub", http.StatusInternalServerError)
+			return
+		}
+		client := github.NewClient(&http.Client{Transport: t})
+
+		if err := processPullRequest(r.Context(), client, &data); err != nil {
+			logrus.Error(err)
+			http.Error(w, "internal server error", http.StatusInternalServerError)
+		}
+		// default:
+		// 	http.Error(w, "unkown kind "+kind, http.StatusBadRequest)
+	}
+}
+
+type comment struct {
+	path    string
+	line    int
+	message string
+}
+
+func processRepo(repo, ref string) ([]comment, error) {
+	logrus.Infof("processing %s:%s", repo, ref)
+	dir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return nil, fmt.Errorf("could not create temp dir: %v", err)
+	}
+	logrus.Debugf("temp dir: %s", dir)
+	defer os.RemoveAll(dir)
+
+	importPath := strings.TrimSuffix(strings.TrimPrefix(repo, "git://"), ".git")
+
+	r := runner{gopath: dir, path: importPath}
+	absPath := filepath.Join(dir, "src", importPath)
+	os.MkdirAll(absPath, os.ModePerm)
+
+	opts := &git.CloneOptions{
+		URL:           repo,
+		ReferenceName: plumbing.ReferenceName(fmt.Sprintf("refs/heads/%s", ref)),
+		SingleBranch:  true,
+	}
+	logrus.Infof("git clone %s -b %s", opts.URL, opts.ReferenceName)
+	repo, err = git.PlainClone(absPath, false, opts)
+	if err != nil {
+		return nil, fmt.Errorf("could not clone repo: %v", err)
+	}
+
+	logrus.Info("fetching dependencies")
+	out, err := r.run("go", "get", ".")
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch dependencies: %v\n%s", err, out)
+	}
+
+	logrus.Info("vetting code")
+	out, err = r.run("megacheck", "./...")
+	logrus.Debugf("megacheck output:\n%s", out)
+	if err == nil {
+		return nil, nil
+	}
+
+	var comments []comment
+
+	s := bufio.NewScanner(bytes.NewReader(out))
+	for s.Scan() {
+		logrus.Infof("processing line: %s", s.Text())
+		ps := strings.SplitN(s.Text(), ":", 4)
+		if len(ps) != 4 {
+			logrus.Errorf("unparsable line %s", s.Text())
+		}
+		path, line, _, msg := ps[0], ps[1], ps[2], ps[3]
```

Comment:
line + column could be used with ASTs to identify the symbol that is commented on